docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,311 @@
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPath
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
+    "/foo/*":
+        service:
+            kind: Service
+            name: foosvc
+        port:
+            name: fooport
+            port: 80
+            targetPort: 8080
+            nodePort: 3230
+    "/bar/*":
+        service:
+            kind: Service
+            name: barsvc
+        port:
+            name: barport
+            port: 80
+            targetPort: 8081
+            nodePort: 3232
+    "/foobar/*":
+        service:
+            kind: Service
+            name: foobarsvc
+        port:
+            name: foobarport
+            port: 80
+            targetPort: 8082
+            nodePort: 3233
+status:
+  # include host here when we have DNS figured out.
+  loadBalancer:
+    ingress:
+    - ip: 104.101.11.313
+```
+
+#### Claims
+
+In the context of 1.1 a cluster will only have one l7 loadbalancer controller that claims all IngressPaths so we don't need claims. In the larger scheme of things, we might still not need claims, however they solve the impedance mismatch problem (why do I need a loadbalancer to expose my service?) because that is how you get an ip. If you want to expose your service, you need a public ip for it. No matter where you're running you can get one by creating a claim. If you want something to use an ip you already have, create a claim for it.
+
+The downside is the user needs to create another resource. We can make this easier by defaulting claim-less ingress paths to a new claim.
```

Comment:
It hasn't come up for us yet except when ops wants to fine tune a
particular case, and in all of those scenarios they can configure the
routers as they need.

On Mon, Aug 17, 2015 at 9:21 PM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/loadbalancing.md
> <https://github.com/kubernetes/kubernetes/pull/12827#discussion_r37255972>
> :
>
> > +            name: foobarport
> > +            port: 80
> > +            targetPort: 8082
> > +            nodePort: 3233
> > +status:
> > +  # include host here when we have DNS figured out.
> > +  loadBalancer:
> > +    ingress:
> > +    - ip: 104.101.11.313
> > +```
> > +
> > +#### Claims
> > +
> > +In the context of 1.1 a cluster will only have one l7 loadbalancer controller that claims all IngressPaths so we don't need claims. In the larger scheme of things, we might still not need claims, however they solve the impedance mismatch problem (why do I need a loadbalancer to expose my service?) because that is how you get an ip. If you want to expose your service, you need a public ip for it. No matter where you're running you can get one by creating a claim. If you want something to use an ip you already have, create a claim for it.
> > +
> > +The downside is the user needs to create another resource. We can make this easier by defaulting claim-less ingress paths to a new claim.
>
> Agreed, I'm not planning on claims before October. Being explicity about
> which ip one wants to join has some advantages but I don't think we need to
> reach a decision right now.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12827/files#r37255972>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift